stl/inc/format

```diff
@@ -445,103 +449,239 @@ _NODISCARD constexpr bool _Is_execution_charset_utf8() {
-    // *_First is not a valid lead byte.
-    _STL_INTERNAL_CHECK(_First < _Last);
+        if (_Ch < 0b1110'0000u) {
+            // check for non-lead byte or partial 2-byte encoded character
+            return (_Ch >= 0b1100'0000u && _Len >= 2) ? 2 : -1;
+        }
 
-    if constexpr (_Is_execution_charset_utf8_v) {
-        return _Utf8_code_units_in_next_character(_First, _Last);
-    } else {
-        switch (_Cvt._Mbcurmax) {
+        if (_Ch < 0b1111'0000u) {
+            // check for partial 3-byte encoded character
+            return (_Len >= 3) ? 3 : -1;
+        }
+
+        // check for partial 4-byte encoded character
+        return (_Len >= 4) ? 4 : -1;
+    }
+
+    _NODISCARD static constexpr int _Estimate_utf8_character_width(const char* const _Ptr, const int _Units) noexcept {
+        // Return an estimate for the width of the character composed of _Units code units,
+        // whose first code unit is denoted by _Ptr.
+        auto _Ch = static_cast<char32_t>(*_Ptr);
+        switch (_Units) {
         default:
-            _STL_INTERNAL_CHECK(!"Bad number of encoding units for this code page");
-            [[fallthrough]];
         case 1:
-            return 1; // all characters have only one code unit
-
         case 2:
-            return _Double_byte_encoding_code_units_in_next_character(_First, _Last, _Cvt);
+            return 1;
+        case 3:
+            _Ch &= 0b1111u;
+            break;
+        case 4:
+            _Ch &= 0b111u;
+            break;
+        }
+
+        for (int _Idx = 1; _Idx < _Units; ++_Idx) {
+            _Ch = _Ch << 6 | (_Ptr[_Idx] & 0b11'1111u);
+        }
 
-        case 4: // Assume UTF-8 (as does _Mbrtowc)
+        if (_Units == 3) {
+            return _Unicode_width_estimate<_Width_estimate_low_intervals>(_Ch);
+        }
+
+        return _Unicode_width_estimate<_Width_estimate_high_intervals>(_Ch);
+    }
+
+public:
+    _NODISCARD int _Units_in_next_character(const char* _First, const char* _Last) const noexcept {
+        // Returns a count of the number of code units that compose the first encoded character in
+        // [_First, _Last), or -1 if [_First, _Last) doesn't contain an entire encoded character or
+        // *_First is not a valid lead byte.
+        _STL_INTERNAL_CHECK(_First < _Last);
+
+        if constexpr (_Statically_Utf8) {
             return _Utf8_code_units_in_next_character(_First, _Last);
+        } else {
+            switch (this->_Cvt._Mbcurmax) {
+            default:
+                _STL_INTERNAL_CHECK(!"Bad number of encoding units for this code page");
+                [[fallthrough]];
+            case 1:
+                return 1; // all characters have only one code unit
+
+            case 2:
+                return this->_Double_byte_encoding_code_units_in_next_character(_First, _Last);
+
+            case 4: // Assume UTF-8 (as does _Mbrtowc)
+                return _Utf8_code_units_in_next_character(_First, _Last);

```

Comment:
Most Standardized conversions to and from wide characters are broken on our implementation because our wide character type (16-bit `wchar_t`) isn't capable of representing every member of the execution wide-character set (Unicode).

Reply:
I thought that was what `_UNICODE` controlled.

Reply:
After in-person discussion, we're happy that this is no worse than the rest of the library's support (not a great bar, I admit) and we'll run with it for now.